toplevel/topdirs.ml

```diff
@@ -549,11 +549,36 @@ let () =
     )
     "Print the signature of the corresponding type constructor."
 
+let is_exception_constructor env type_expr =
+  Ctype.equal env true [type_expr] [Predef.type_exn]
+
+let () =
+  reg_show_prim "show_non_exception_constructor"
+    (fun env loc _id lid ->
+       let desc = Env.lookup_constructor ~loc Env.Positive lid env in
+       if is_exception_constructor env desc.cstr_res then
+         raise Not_found;
+       let path =
+         match Ctype.repr desc.cstr_res with
+         | {desc=Tconstr(path, _, _)} -> path
+         | _ -> raise Not_found
+       in
+       let type_decl = Env.find_type path env in
+       let type_id =
+         match path with
+         | Pident type_id -> type_id
+         | Pdot (_, s) -> Ident.create_local s
+         | Papply _ -> raise Not_found
+       in
+       [ Sig_type (type_id, type_decl, Trec_not, Exported) ]
```

Comment:
Unconditionally using `Trec_not` here will print type definitions with `nonrec`, even if the types are defined recursively.

Reply:
thanks, I'll look into how to fix that too.  Looks like I can create another bug report for show_type as that unconditionally uses Trec_not too?

Reply:
> Looks like I can create another bug report for show_type as that unconditionally uses `Trec_not` too?

That's a good idea (assuming there isn't one already).

Reply:
So my current plan of attack for this bit is to try and extract out the rec_flag (Recursive/Nonrecursive) from the Parsetree.attributes.attr_payload of the constructor_description.cstr_attributes.

Is that the right way to go or am I missing a trick?

My next question is then how to traverse around on the parsetree...

Reply:
also can I just clarify what we'd like to see for this last nonrec piece.  For something like:
```
type foo = Foo of bar
and bar = Bar of foo
```

we'd want to see
```
# #show Foo;;
type rec foo = Foo of bar
```

then people can dig further with showing foo/bar/Bar etc. rather than listing out all the and cases at once).

Obviously for non-recursive things it will work as is:
```
# #show Some
type nonrec 'a option = None | Some of 'a
```

Reply:
and I think I am missing a trick with the parsetree.attributes.

looking more it seems that a good starting point would be Typedecl.check_recursion function and what is going on in that?

Reply:
You don't need to fix this issue in this PR. I think it is fine to keep `Trec_not` for now, in order to be consistent with t`show_type`.  Another option would be to switch to Trec_first` for both. Printing with the right recursivity qualifier is a subject for another PR, and one quite more complex than this one.

Reply:
It looks unreasonably difficult to me to access this information from a `Types.type_declaration` value. I'm not super-familiar with this part of the codebase, but it looks like `Types.type_declaration` lacks a lot of useful surface-level information compared to what is available in the `Typedtree` (and in fact ocamldoc reimplements its own environment-gathering logic from the full typedtree in `Odoc_ast.Typedtree_search`.

In your stead I would cut my losses by just using `Trec_first` instead of `Trec_not` bluntly, which *should* result in `nonrec` not being printed. It's a safer default.